### PR TITLE
WIP: Fix lack of D1 typedef support in `ocean.meta.types.ReduceType`

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Will be needed to switch deserializer to use it

### CODE REVIEW

File: src/ocean/meta/traits/Basic.d [hunk 1]
```diff
@@ -546,3 +546,51 @@ unittest
...
+
+    Template_Params:
+        T = type to check
+
+    Evaluates to:
+        true if T is a typedef, false otherwise
+
+*******************************************************************************/
+
+version (D_Version2)
```
Comment (mathias-lang-sociomantic) [inline src/ocean/meta/traits/Basic.d:566]: Why can't the version be inside the template ?

File: src/ocean/meta/traits/Basic.d [hunk 2]
```diff
@@ -546,3 +546,51 @@ unittest
...
+        }
+        else
+        {
+            const bool isD1Typedef = false;
+        }
+    }
+
+    unittest
+    {
+        typedef double RealNum;
```
Comment (mathias-lang-sociomantic) [inline src/ocean/meta/traits/Basic.d:590]: Likewise, I would use the ocean helper. No one use `typedef` directly nowadays anyway.

File: src/ocean/meta/traits/Basic.d [hunk 3]
```diff
@@ -546,3 +546,51 @@ unittest
...
+
+/*******************************************************************************
+
+    Determines if T is a D1 typedef.
+
+    Typedef has been removed in D2 and this template will always evaluate to
+    false if compiled with version = D_Version2. Once ocean drops D1 support,
+    all references to this template can be removed.
+
+    Template_Params:
```
Comment (mathias-lang-sociomantic) [inline src/ocean/meta/traits/Basic.d:558]: Params

**Comment by mathias-lang-sociomantic** [comment]
You know what I think about this :)

**Comment by mihails-strasuns-sociomantic** [comment]
> You know what I think about this :)

Yes, and I am not yet sure this is a final version. Ideally I want to reduce implementation of both serializer traits and `TypeId` to the same core set of primitives but those have somewhat conflicting requirements. I keep changing my mind back and forth as I work on it.

Adjusted title to make obvious it is a raw version.

**Comment by mihails-strasuns-sociomantic** [comment]
Closing in favor of https://github.com/sociomantic-tsunami/ocean/pull/178

**Pull request closed without merging**